connect-inject/health_check_resource_test.go

```diff
@@ -0,0 +1,551 @@
+						annotationService: testServiceNameAnnotation,
+					},
+				},
+				Spec: corev1.PodSpec{
+					Containers: []corev1.Container{
+						corev1.Container{
+							Name: testPodName,
+						},
+					},
+				},
+				Status: corev1.PodStatus{
+					HostIP: "127.0.0.1",
+					Phase:  corev1.PodRunning,
+					Conditions: []corev1.PodCondition{{
+						Type:   corev1.PodReady,
+						Status: corev1.ConditionFalse,
+					}},
+				},
+			},
+			&api.AgentCheck{
+				CheckID: testHealthCheckID,
+				Status:  api.HealthCritical,
+			},
+			"",
+		},
+		{
+			"Reconcile existing object from failing to passing",
+			true,
+			api.HealthCritical,
+			&corev1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      testPodName,
+					Namespace: "default",
+					Labels:    map[string]string{labelInject: "true"},
+					Annotations: map[string]string{
+						annotationStatus:  "injected",
+						annotationService: testServiceNameAnnotation,
+					},
+				},
+				Spec: corev1.PodSpec{
+					Containers: []corev1.Container{
+						corev1.Container{
+							Name: testPodName,
+						},
+					},
+				},
+				Status: corev1.PodStatus{
+					HostIP: "127.0.0.1",
+					Phase:  corev1.PodRunning,
+					Conditions: []corev1.PodCondition{{
+						Type:   corev1.PodReady,
+						Status: corev1.ConditionTrue,
+					}},
+				},
+			},
+			&api.AgentCheck{
+				CheckID: testHealthCheckID,
+				Status:  api.HealthPassing,
+			},
+			"",
+		},
+		{
+			"Reconcile pod not running no update",
+			false,
+			api.HealthCritical,
+			&corev1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      testPodName,
+					Namespace: "default",
+					Labels:    map[string]string{labelInject: "true"},
+					Annotations: map[string]string{
+						annotationStatus:  "injected",
+						annotationService: testServiceNameAnnotation,
+					},
+				},
+				Spec: corev1.PodSpec{
+					Containers: []corev1.Container{
+						corev1.Container{
+							Name: testPodName,
+						},
+					},
+				},
+				Status: corev1.PodStatus{
+					HostIP: "127.0.0.1",
+					Phase:  corev1.PodFailed,
+					Conditions: []corev1.PodCondition{{
+						Type:   corev1.PodReady,
+						Status: corev1.ConditionFalse,
+					}},
+				},
+			},
+			nil,
+			"",
+		},
+	}
+	for _, tt := range cases {
+		t.Run(tt.Name, func(t *testing.T) {
+			require := require.New(t)
+
+			server, client, resource := testServerAgentResourceAndController(t, tt.Pod)
+			defer server.Stop()
+
+			// Create a passing service
+			server.AddService(t, testServiceNameReg, "passing", nil)
+			if tt.CreateHealthCheck {
+				// register the health check if ObjectCreate didnt run
+				err := registerHealthCheck(t, client, tt.InitialState)
+				require.NoError(err)
+			}
+
+			err := resource.Reconcile()
+			require.NoError(err)
+			actual := testGetConsulAgentChecks(t, client)
+			if tt.Expected == nil || actual == nil {
+				require.Equal(tt.Expected, actual)
+			} else {
+				require.Equal(tt.Expected.Status, actual.Status)
+			}
```

Comment:
Why do we need this check? We're never passing `tt.Expected` that is `nil` I think this should be:
```suggestion
			require.NotNil(actual)
			require.Equal(tt.Expected, actual)
```

Reply:
actual can be nil in the (negative) cases where the check isn't created and the pod not processed.

Reply:
Oh ok, in that case should it be just:
```suggestion
				require.Equal(tt.Expected, actual)
```

Not sure why you're ignoring all other fields in the case when both expected and actual are not nil?